TiFlash results materialization (#11701)
pingcap/docs-cn: tiflash/tiflash-results-materialization.md

---
title: TiFlash 计算结果物化

> summary: 介绍如何在同一个事务处理中物化保存 TiFlash 的计算结果。本功能为实验功能。

Reviewer: ```suggestion
summary: 介绍如何在同一个事务处理中物化保存 TiFlash 的计算结果。
```

---

# TiFlash 计算结果物化

> 本文介绍如何在同一个事务处理中实现 TiFlash 计算结果的物化保存。

Reviewer: ```suggestion
本文介绍如何在同一个事务处理中实现 TiFlash 计算结果的物化保存。

> **警告：**
>
> 当前该功能为实验特性，不建议在生产环境中使用。
```

## 支持下推的算子

TiFlash 查询的结果可以通过 `INSERT INTO SELECT` 语句保存在一个 TiDB 表中：

```sql
INSERT [LOW_PRIORITY | HIGH_PRIORITY] [IGNORE]
    [INTO] tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [(col_name [, col_name] ...)]
    SELECT ...
    [ON DUPLICATE KEY UPDATE assignment_list]value:
    {expr | DEFAULT}

assignment:
    col_name = valueassignment_list:
    assignment [, assignment] ...
```

## 使用场景和限制

### 执行过程

* SELECT 子句在 TiFlash 上执行，一般情况下优化器会自动选择，需要用户手动干预时可以通过设置 "tidb_enforce_mpp = TRUE" 来强制查询计划走 TiFlash
* SELECT 子句返回的结果首先回到集群中某单一 TiDB Server 节点，然后写入目标表（可以有 TiFlash 副本）
* 整条语句的执行保证 ACID 特性

### 限制

> * 对写入部分事务大小（SELECT 子句返回的结果集）的硬性限制为 1 GB，推荐的使用场景是 100 MB以下。

Reviewer 2: 这里可以引用一下 https://docs.pingcap.com/zh/tidb/dev/transaction-overview#%E4%BA%8B%E5%8A%A1%E9%99%90%E5%88%B6？
这里的逻辑跟普通写入事务是一样的。默认事务大小是 100MB，要更大得调整 `txn-total-size-limit`.

Reviewer 2: 事务太大还有另外一种方法是 https://docs.pingcap.com/zh/tidb/dev/system-variables#tidb_enable_batch_dml
把大事务拆成多个小事务，我理解这个也是值得一提的？如果出错用户可以删表重新来。

---
title: TiFlash 查询结果物化
summary: 介绍如何在同一个事务中保存 TiFlash 的计算结果。
---

# TiFlash 计算结果物化

> **警告：**
>

> > 该功能目前是实验性功能，其形式和使用方法可能会在未来版本中发生变化。

Reviewer 3: @moyun PTAL

> > 该功能目前是实验性功能，不建议在生产环境中使用，其形式和使用方法可能会在未来版本中发生变化。

Author: 该功能目前是实验性功能，请注意使用场景限制。该功能会在未事先通知的情况下发生变化或删除。语法和实现可能会在 GA 前发生变化。如果发现 bug，请提 [Issues · pingcap/tidb](https://github.com/pingcap/tidb/issues) 反馈。

这段警告语的标准格式反复变化是不行的，需要确定下来。

Reviewer 3: 我记录到 feature 文档模板中了，方便以后引用 https://github.com/pingcap/docs-cn/pull/12177/files

> > 该功能目前是实验性功能，不建议在生产环境中使用，其形式和使用方法可能会在未来版本中发生变化。

Reviewer 3: ```suggestion
> 该功能目前是实验性功能，请注意使用场景限制。该功能会在未事先通知的情况下发生变化或删除。语法和实现可能会在 GA 前发生变化。如果发现 bug，请提 [Issues · pingcap/tidb](https://github.com/pingcap/tidb/issues) 反馈。
```

本文介绍如何在同一个事务 (`INSERT INTO SELECT`) 中实现 TiFlash 计算结果物化至某一指定的 TiDB 表中。

> 从 v6.4.0 起，执行 `INSERT INTO SELECT` 语句时，通过将 SELECT 子句下推到 TiFlash 可以把 TiFlash 计算得到的查询结果保存到指定的 TiDB 表中，即物化了 TiFlash 的查询结果。v6.4.0 之前的 TiDB 版本不允许此类行为，即通过 TiFlash 执行的查询必须是只读的，用户需要从应用程序层面接收 TiFlash 返回的结果，然后另行在其它事务或处理中保存结果。

Reviewer 3: ```suggestion
从 v6.5.0 起，执行 `INSERT INTO SELECT` 语句时，通过将 SELECT 子句下推到 TiFlash 可以把 TiFlash 计算得到的查询结果保存到指定的 TiDB 表中，即物化了 TiFlash 的查询结果。v6.5.0 之前的 TiDB 版本不允许此类行为，即通过 TiFlash 执行的查询必须是只读的，用户需要从应用程序层面接收 TiFlash 返回的结果，然后另行在其它事务或处理中保存结果。
```

> **注意：**
>

> > 默认情况下 ([`tidb_allow_mpp = ON`](/system-variables#tidb_allow_mpp-从-v50-版本开始引入))，TiDB 优化器将依据查询代价智能选择下推查询到 TiKV 或 TiFlash。如需强制使用 TiFlash 查询，你可以设置系统变量 [`tidb_enforce_mpp = ON`](/system-variables#tidb_enforce_mpp-从-v51-版本开始引入)。在实验功能阶段，为了使 `tidb_enforce_mpp = ON` 时不影响 其他既存的读写混合事务，特别增加了一个单独控制此项功能的系统变量 [`tidb_enable_tiflash_read_for_write_stmt`](/system-variables#tidb_enable_tiflash_read_for_write_stmt)，用户须设置此变量为 `TRUE` 方可开启本功能。

Reviewer 3: ```suggestion
> - 默认情况下 ([`tidb_allow_mpp = ON`](/system-variables#tidb_allow_mpp-从-v50-版本开始引入))，TiDB 优化器将依据查询代价智能选择下推查询到 TiKV 或 TiFlash。如需强制使用 TiFlash 查询，你可以设置系统变量 [`tidb_enforce_mpp = ON`](/system-variables#tidb_enforce_mpp-从-v51-版本开始引入)。
> - 在实验功能阶段，为了使 `tidb_enforce_mpp = ON` 不影响其他现有的读写混合事务，TiFlash 计算结果物化功能默认关闭。要开启此功能，请设置 [`tidb_enable_tiflash_read_for_write_stmt`](/system-variables#tidb_enable_tiflash_read_for_write_stmt) 为 `ON` 。
```

`INSERT INTO SELECT` 语法如下：

```sql
INSERT [LOW_PRIORITY | HIGH_PRIORITY] [IGNORE]
    [INTO] tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [(col_name [, col_name] ...)]
    SELECT ...
    [ON DUPLICATE KEY UPDATE assignment_list]value:
    {expr | DEFAULT}

assignment:
    col_name = valueassignment_list:
    assignment [, assignment] ...
```

例如，通过以下 `INSERT INTO SELECT` 语句，你可以将 SELECT 子句中表 `t1` 的查询结果保存到表 `t2` 中：

```sql
INSERT INTO t2 (name, country)
SELECT app_name, country FROM t1;
```

## 典型和推荐的使用场景

- 高效的 BI 解决方案

    很多报表类应用有较重的分析查询，如果有很多用户同时打开和刷新报表，则会产生较多的查询请求。一个有效的解决方案是使用本功能在 TiDB 表中保存报表查询的结果，报表刷新时再从结果表中抽取数据，则可以避免多次重复的分析计算。同理，在保存历史分析记录的基础上，可以进一步优化长时间历史数据分析的计算量。例如，某报表 A 用于分析每日的销售利润，使用本功能你可以将报表 A 中每日的分析结果保存到某结果表 T 中。那么，在生成报表 B 分析过去一个月的销售利润时，可以直接使用表 T 中的每日分析结果数据，不仅大幅降低计算量也提升了查询响应速度，减轻系统负载。

- 使用 TiFlash 服务在线应用

    TiFlash 支持的并发请求数量视数据量和查询复杂度不同，但一般不会超过 100 QPS。你可以使用本功能保存 TiFlash 的查询结果，然后通过查询结果表来支持在线的高并发请求。后台的结果表数据更新可以以较低的频率进行，例如以 0.5 秒间隔更新结果表数据也远低于 TiFlash 的并发上限，同时仍然较好地保证了数据新鲜度。

## 执行过程

* 在 `INSERT INTO SELECT` 语句的执行过程中，TiFlash 首先将 SELECT 子句的查询结果返回到集群中某单一 TiDB server 节点，然后再写入目标表（可以有 TiFlash 副本）。
* `INSERT INTO SELECT` 语句的执行保证 ACID 特性。

## 限制

> * TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的可以通过设定 TiDB 控制事务大小的变量 `performance.txn-total-size-limit` 来进行，推荐的使用场景是 100 MiB 以下。

Reviewer 3: ```suggestion
* TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的限制与 TiDB  对单个事务大小的限制相同，可以通过 [performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit) 配置项调整该限制，推荐的使用场景是 100 MiB 以下。
```

> * TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的限制即为 TiDB  对单个事务大小的限制，可以通过 [performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit) 配置项调整该限制，推荐的使用场景是 100 MiB 以下。

Reviewer 3: ```suggestion
* TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的限制即为 TiDB  对单个事务大小的限制，可以通过 [`performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit`) 配置项调整该限制，推荐的使用场景是 100 MiB 以下。
```

>     若 SELECT 返回结果大小超过了事务大小限制的阈值，那么整条语句将会被强制终止并返回错误信息（参见 `performance.txn-total-size-limit`）。

Reviewer 3: ```suggestion
    若 SELECT 返回结果大小超过了事务大小限制的阈值，那么整条语句将会被强制终止并返回错误信息（参见 [performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit)）。
```

---
title: TiFlash 查询结果物化
summary: 介绍如何在同一个事务中保存 TiFlash 的计算结果。
---

# TiFlash 计算结果物化

> **警告：**
>

> > 该功能目前是实验性功能，请注意使用场景限制。该功能会在未事先通知的情况下发生变化或删除。语法和实现可能会在 GA 前发生变化。如果发现 bug，请提 [Issues · pingcap/tidb](https://github.com/pingcap/tidb/issues) 反馈。

Reviewer 3: 嗯嗯，更清楚了，谢谢，对了，“反馈”两个字可以留着吗

> 本文介绍如何在同一个事务 (`INSERT INTO SELECT`) 中实现 TiFlash 查询结果物化至某一指定的 TiDB 表中。
> 
> 从 v6.5.0 起，执行 `INSERT INTO SELECT` 语句时，通过将 `SELECT` 子句下推到 TiFlash 可以把 TiFlash 计算得到的查询结果保存到指定的 TiDB 表中，即物化了 TiFlash 的查询结果。v6.5.0 之前的 TiDB 版本不允许此类行为，即通过 TiFlash 执行的查询必须是只读的，你需要从应用程序层面接收 TiFlash 返回的结果，然后另行在其它事务或处理中保存结果。

Reviewer 3: ```suggestion
本文介绍如何在同一个事务 (`INSERT INTO SELECT`) 中实现将 TiFlash 查询结果保存至某一指定的 TiDB 表中。

从 v6.5.0 起，TiDB 支持将 TiFlash 查询结果保存到数据表中，即物化了 TiFlash 的查询结果。执行 `INSERT INTO SELECT` 语句时，如果 TiDB 将 `SELECT` 子查询下推到了 TiFlash， TiFlash 的查询结果可以保存到 `INSERT INTO` 指定的 TiDB 表中。v6.5.0 之前的 TiDB 版本不允许此类行为，即通过 TiFlash 执行的查询必须是只读的，你需要从应用程序层面接收 TiFlash 返回的结果，然后另行在其它事务或处理中保存结果。
```

> 从 v6.5.0 起，TiDB 支持将 TiFlash 查询结果保存到数据表中，即物化了 TiFlash 的查询结果。执行 `INSERT INTO SELECT` 语句时，如果 TiDB 将 `SELECT` 子查询下推到了 TiFlash， TiFlash 的查询结果可以保存到 `INSERT INTO` 指定的 TiDB 表中。v6.5.0 之前的 TiDB 版本不允许此类行为，即通过 TiFlash 执行的查询必须是只读的，你需要从应用程序层面接收 TiFlash 返回的结果，然后另行在其它事务或处理中保存结果。

Reviewer 3: ```suggestion
从 v6.5.0 起，TiDB 支持将 TiFlash 查询结果保存到数据表中，即物化了 TiFlash 的查询结果。执行 `INSERT INTO SELECT` 语句时，如果 TiDB 将 `SELECT` 子查询下推到了 TiFlash，TiFlash 的查询结果可以保存到 `INSERT INTO` 指定的 TiDB 表中。v6.5.0 之前的 TiDB 版本不允许此类行为，即通过 TiFlash 执行的查询必须是只读的，你需要从应用程序层面接收 TiFlash 返回的结果，然后另行在其它事务或处理中保存结果。
```

> **注意：**
>

> > - 默认情况下 ([`tidb_allow_mpp = ON`](/system-variables#tidb_allow_mpp-从-v50-版本开始引入))，TiDB 优化器将依据查询代价智能选择下推查询到 TiKV 或 TiFlash。如需强制使用 TiFlash 查询，你可以设置系统变量 [`tidb_enforce_mpp`](/system-variables#tidb_enforce_mpp-从-v51-版本开始引入) 为 `ON`。

Reviewer 3: ```suggestion
> - 默认情况下 ([`tidb_allow_mpp = ON`](/system-variables#tidb_allow_mpp-从-v50-版本开始引入))，TiDB 优化器将依据查询代价智能选择下推查询到 TiKV 或 TiFlash。如需强制使用 TiFlash 查询，你可以设置系统变量 [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-从-v51-版本开始引入) 为 `ON`。
```

> > - 默认情况下 ([`tidb_allow_mpp = ON`](/system-variables#tidb_allow_mpp-从-v50-版本开始引入))，TiDB 优化器将依据查询代价智能选择下推查询到 TiKV 或 TiFlash。如需强制使用 TiFlash 查询，你可以设置系统变量 [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-从-v51-版本开始引入) 为 `ON`。

Reviewer 3: ```suggestion
> - 默认情况下 ([`tidb_allow_mpp = ON`](/system-variables.md#tidb_allow_mpp-从-v50-版本开始引入))，TiDB 优化器将依据查询代价智能选择下推查询到 TiKV 或 TiFlash。如需强制使用 TiFlash 查询，你可以设置系统变量 [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-从-v51-版本开始引入) 为 `ON`。
```

> > - 在实验特性阶段，该功能默认关闭。要开启此功能，请设置系统变量 [`tidb_enable_tiflash_read_for_write_stmt`](/system-variables#tidb_enable_tiflash_read_for_write_stmt) 为 `ON`。

Reviewer 3: ```suggestion
> - 在实验特性阶段，该功能默认关闭。要开启此功能，请设置系统变量 [`tidb_enable_tiflash_read_for_write_stmt`](/system-variables#tidb_enable_tiflash_read_for_write_stmt-从-v630-版本开始引入) 为 `ON`。
```

> > - 在实验特性阶段，该功能默认关闭。要开启此功能，请设置系统变量 [`tidb_enable_tiflash_read_for_write_stmt`](/system-variables#tidb_enable_tiflash_read_for_write_stmt-从-v630-版本开始引入) 为 `ON`。

Reviewer 3: ```suggestion
> - 在实验特性阶段，该功能默认关闭。要开启此功能，请设置系统变量 [`tidb_enable_tiflash_read_for_write_stmt`](/system-variables#tidb_enable_tiflash_read_for_write_stmt-从-v630-版本开始引入) 为 `ON`。
```

`INSERT INTO SELECT` 语法如下：

```sql
INSERT [LOW_PRIORITY | HIGH_PRIORITY] [IGNORE]
    [INTO] tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [(col_name [, col_name] ...)]
    SELECT ...
    [ON DUPLICATE KEY UPDATE assignment_list]value:
    {expr | DEFAULT}

assignment:
    col_name = valueassignment_list:
    assignment [, assignment] ...
```

例如，通过以下 `INSERT INTO SELECT` 语句，你可以将 `SELECT` 子句中表 `t1` 的查询结果保存到表 `t2` 中：

```sql
INSERT INTO t2 (name, country)
SELECT app_name, country FROM t1;
```

## 典型和推荐的使用场景

- 高效的 BI 解决方案

    很多报表类应用有较重的分析查询，如果有很多用户同时打开和刷新报表，则会产生较多的查询请求。一个有效的解决方案是使用本功能在 TiDB 表中保存报表查询的结果，报表刷新时再从结果表中抽取数据，则可以避免多次重复的分析计算。同理，在保存历史分析记录的基础上，可以进一步优化长时间历史数据分析的计算量。例如，某报表 A 用于分析每日的销售利润，使用本功能你可以将报表 A 中每日的分析结果保存到某结果表 T 中。那么，在生成报表 B 分析过去一个月的销售利润时，可以直接使用表 T 中的每日分析结果数据，不仅大幅降低计算量也提升了查询响应速度，减轻系统负载。

- 使用 TiFlash 服务在线应用

    TiFlash 支持的并发请求数量视数据量和查询复杂度不同，但一般不会超过 100 QPS。你可以使用本功能保存 TiFlash 的查询结果，然后通过查询结果表来支持在线的高并发请求。后台的结果表数据更新可以以较低的频率进行，例如以 0.5 秒间隔更新结果表数据也远低于 TiFlash 的并发上限，同时仍然较好地保证了数据新鲜度。

## 执行过程

* 在 `INSERT INTO SELECT` 语句的执行过程中，TiFlash 首先将 `SELECT` 子句的查询结果返回到集群中某单一 TiDB server 节点，然后再写入目标表（可以有 TiFlash 副本）。
* `INSERT INTO SELECT` 语句的执行保证 ACID 特性。

## 限制

> * TiDB 对 `INSERT INTO SELECT` 语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables#tidb_mem_quota_query) 调整。

Reviewer 3: ```suggestion
* TiDB 对 `INSERT INTO SELECT` 语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 调整。
```

> * TiDB 对 `INSERT INTO SELECT` 语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 调整。

Reviewer 2: ```suggestion
* TiDB 对 `INSERT INTO SELECT` 语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 调整。从 v6.5.0 版本开始，不推荐使用 [txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit) 来控制事务内存大小，详见该配置项文档。
```

>     更多信息，请参考[用户文档](/configure-memory-usage.md)。

Reviewer 3: ```suggestion
    更多信息，请参考[TiDB 内存控制文档](/configure-memory-usage.md)。
```

> * TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的限制即为 TiDB  对单个事务大小的限制，可以通过 [`performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit`) 配置项调整该限制，推荐的使用场景是 100 MiB 以下。

Reviewer 3: ```suggestion
* TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的限制即为 TiDB 对单个事务大小的限制，可以通过 [`performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit`) 配置项调整该限制，推荐的使用场景是 100 MiB 以下。
```

> * TiDB 对 `INSERT INTO SELECT` 语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 调整。从 v6.5.0 版本开始，不推荐使用 [txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit) 来控制事务内存大小，详见该配置项文档。

Reviewer 3: ```suggestion
* TiDB 对 `INSERT INTO SELECT` 语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 调整。从 v6.5.0 版本开始，不推荐使用 [`txn-total-size-limit`](/tidb-configuration-file.md#txn-total-size-limit) 来控制事务内存大小，详见该配置项文档。
```

> * TiDB 对 `SELECT` 子句返回的结果集（即 `INSERT` 写入的事务）大小的限制即为 TiDB 对单个事务大小的限制，可以通过 [`performance.txn-total-size-limit](/tidb-configuration-file.md#txn-total-size-limit`) 配置项调整该限制，推荐的使用场景是 100 MiB 以下。
> 
>     若 `SELECT` 返回结果大小超过了事务大小限制的阈值 [`performance.txn-total-size-limit`](/tidb-configuration-file.md#txn-total-size-limit)，那么整条语句将会被强制终止并返回错误信息。

Reviewer 2: ```suggestion
* TiDB 对该语句的内存限制可以通过系统变量 [`tidb_mem_quota_query`](/system-variables.md#tidbmemquotaquery) 调整。

    更多信息，请参考[用户文档](/configure-memory-usage.md)。
```

    col_name = valueassignment_list:
    assignment [, assignment] ...
```

例如，通过以下 `INSERT INTO SELECT` 语句，你可以将 SELECT 子句中表 `t1` 的查询结果保存到表 `t2` 中：

```sql
INSERT INTO t2 (name, country)
SELECT app_name, country FROM t1;
```

## 典型和推荐的使用场景

- 高效的 BI 解决方案

  很多报表类应用有较重的分析查询，如果有很多用户在同时打开和刷新报表，则会产生较多的查询请求。一个有效的解决方案是使用本功能在 TiDB 表中保存报表查询的结果，报表刷新时再从结果表中抽取数据，则可以避免多次重复的分析计算。同理，在保存历史分析记录的基础上，可以进一步优化长时间历史数据分析的计算量。例如，某报表 A 用于分析每日的销售利润，使用本功能你可以将报表 A 中每日的分析结果保存到某结果表 T 中。那么，在生成报表 B 分析过去一月窗口的销售利润时，可以直接使用表 T 中的每日分析结果数据，不仅大幅降低计算量也提升了查询响应速度，减轻系统负载。

- 使用 TiFlash 服务在线应用

  TiFlash 支持的并发请求数量视数据量和查询复杂度不同，但一般不会超过 100 QPS。你可以使用本功能保存 TiFlash 的查询结果，然后通过查询结果表来支持在线的高并发请求。后台的结果表数据更新可以以较低的频率进行，例如以 0.5s 间隔更新结果表数据也远低于 TiFlash 的并发上限，同时仍然较好地保证了数据新鲜度。

## 执行过程

* 在`INSERT INTO SELECT` 语句的执行过程中，TiFlash 首先将 SELECT 子句的查询结果返回到集群中某单一 TiDB Server 节点，然后再写入目标表（可以有 TiFlash 副本）。
* `INSERT INTO SELECT` 语句的执行保证 ACID 特性。

## 限制

* TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的硬性限制为 1 GB，推荐的使用场景是 100 MB 以下。

    若 SELECT 返回结果大小超过了 1 GB，那么整条语句将会被强制终止。用户会得到以下出错信息:

    `The query produced a too large intermediate result and thus failed`

* TiDB 对 `INSERT INTO SELECT` 语句的并发没有硬性限制，但是推荐考虑以下用法：

    * 当“写事务”较大时，例如接近 1 GB， 建议控制并发不超过 10。

>     * 当“写事务”较小时，例如小于 100 MB， 建议控制并发不超过 30。

Reviewer 2: 限制这个还是没有改，根据事务文档 https://docs.pingcap.com/zh/tidb/stable/transaction-overview#%E4%BA%8B%E5%8A%A1%E9%99%90%E5%88%B6。
> TiDB 中，单个事务的总大小默认不超过 100 MB，这个默认值可以通过配置文件中的配置项 txn-total-size-limit 进行修改，最大支持 1 TB。单个事务的实际大小限制还取决于服务器剩余可用内存的大小，执行事务时 TiDB 进程的内存消耗相对于事务大小会存在一定程度的放大，最大可能达到提交事务大小的 2 到 3 倍以上。

Reviewer 3: @gengliqi 你建议这里如何修改下呢

Reviewer 2: > * TiDB 对 SELECT 子句返回的结果集（即 INSERT 写入的事务）大小的硬性限制为 1 GB，推荐的使用场景是 100 MB 以下。
>
>    若 SELECT 返回结果大小超过了 1 GB，那么整条语句将会被强制终止。用户会得到以下出错信息:
>
>    `The query produced a too large intermediate result and thus failed`
我建议这段直接引用事务的文档，因为这里逻辑是一样的，没有专门针对 insert select 的限制。

Reviewer 3: @moyun PTAL

Author: > 限制这个还是没有改，根据事务文档 [https://docs.pingcap.com/zh/tidb/stable/transaction-overview#%E4%BA%8B%E5%8A%A1%E9%99%90%E5%88%B6。](https://docs.pingcap.com/zh/tidb/stable/transaction-overview#%E4%BA%8B%E5%8A%A1%E9%99%90%E5%88%B6%E3%80%82)
> 
> > TiDB 中，单个事务的总大小默认不超过 100 MB，这个默认值可以通过配置文件中的配置项 txn-total-size-limit 进行修改，最大支持 1 TB。单个事务的实际大小限制还取决于服务器剩余可用内存的大小，执行事务时 TiDB 进程的内存消耗相对于事务大小会存在一定程度的放大，最大可能达到提交事务大小的 2 到 3 倍以上。

there is no conflict here with the tidb doc, and the conclusion here is specifically for this scenario but just happenly same as the tidb doc.
